src/mantine-core/src/components/NumberInput/NumberInput.tsx

```diff
@@ -175,6 +185,52 @@ export const NumberInput = forwardRef<HTMLInputElement, NumberInputProps>(
       }
     }, [value]);
 
+    const shouldUseStepInterval =
+      stepIncrementInitialDelay !== undefined && stepIncrementInterval !== undefined;
+
+    const onStepTimeoutRef = useRef<NodeJS.Timeout>(null);
+
+    const onStepDone = () => {
+      if (onStepTimeoutRef.current) {
+        clearTimeout(onStepTimeoutRef.current);
+        onStepTimeoutRef.current = null;
+      }
+    };
+
+    const onStepHandleChange = (isIncrement: boolean) => {
+      if (isIncrement) {
+        incrementRef.current();
+      } else {
+        decrementRef.current();
+      }
+    };
+
+    const onStepLoop = (isIncrement: boolean) => {
+      onStepHandleChange(isIncrement);
+
+      if (shouldUseStepInterval) {
+        onStepTimeoutRef.current = setTimeout(() => onStepLoop(isIncrement), stepIncrementInterval);
```

Comment:
To make `setTimeout` work with `number` type, replace it with `window.setTimeout`, same applies to `clearTimeout`